### DISCUSSION TITLE
[RELEASED] BottomSheet v4

### DISCUSSION THREAD

**Discussion by gorhom** [gorhom/react-native-bottom-sheet#427] [Ideas]
# Overview
Hi everyone 👋

After introducing [Keyboard Handling](https://github.com/gorhom/react-native-bottom-sheet/discussions/233) on the previous discussion, I notice that the core library needs a refactor especially when it comes to state and memoized variables. Some of the issues that needed this major refactor https://github.com/gorhom/react-native-bottom-sheet/issues/401 https://github.com/gorhom/react-native-bottom-sheet/issues/402, and others.
 
Another area I wanted to improve is logging and in-code docs to help the community to contribute and debug issues easily, so I will try to document the major functionalities and main components.

Finally, I would love to hear your feedback or improvements that could be shipped with this release.

## Testing

I will be posting an individual PRs for each feature/improvements and include installation commands.

You can test latest alpha here:

```bash
yarn add @gorhom/bottom-sheet@4.0.0-alpha.30
```

Any issue should be reported with this reproducible issue template on Expo Snack
https://snack.expo.io/@gorhom/bottom-sheet-v4-reproducible-issue-template

## Status

We've reach to the sable version so far, I am currently working on finalising this release.

- [x] Rename branch `Master` to `v2`
- [x] Rename branch `v4` to `Master`
- [x] Update the documentation website 

#### New Features
- [x] Added custom pan gesture and scroll handler customisation support (#525) (by @vonovak)
- [x] Added `snapToPosition` method which will allows user to snap to a custom position out of the provided snap points.
- [x] Added `bottomInset` prop which will add a space below the `BottomSheet`.
- [x] Added `enablePanDownToClose` prop which will allow user to close the sheet when panning down.
- [x] Added `footerComponent` prop.
- [x] Added `BottomSheetVirtualizedList`
- [x] Added keyboard input mode for Android.
- [x] Added Pull-To-Refresh support.
- [x] Added Detached BottomSheet 
- [x] Added `backgroundStyle`, `handleStyle` & `handleIndicatorStyle` to the `BottomSheet`
- [ ] ~~Added `Web` support. TBD~~

#### Improvements
- [x] Converted all internal state/memoized variables to Reanimated shared values.
- [x] Prevent resetting the scrollable position when dragging by the handle. #461 
- [x] Added `animatedPosition` & `animatedIndex` to `useBottomSheet`.
- [x] Export `useBottomSheetInternal` for advanced usage.
- [x] Improved dynamic snap points handling, and introduce `useBottomSheetDynamicSnapPoints` hook
- **TBD**

#### Breaking Changes
- [x] Removed `animationEasing` from `BottomSheet` props.
- [x] Removed `animationDuration` from `BottomSheet` props.
- [x] Removed `closeOnPress` from `BottomSheetBackdrop` props.
- [x] Removed `dismissOnClose` from `BottomSheetModal` props.

#### DevX
- [ ] Added visuals of the BottomSheet layering to the documentation.
- [x] Added `enableLogging` to enable logging in development environment.
- [x] Added more loggings.

**Comment by gorhom** [comment]
## Converted all internal state/memoized variables to Reanimated shared values.

In this branch, I have refactored the library to use shared values to store and update layout values.

### Notice

Calculating snap points position had change in this branch, now the library will count the handle height when positioning the sheet. This is a **breaking change** and will require developers to carefully monitor the change.

This is how you can visualise it: 

![snap-point-calculation](https://user-images.githubusercontent.com/4061838/117367506-2c945200-aeba-11eb-86dc-4218b41ac42a.jpg)

### Testing

```bash
yarn add ssh://git@github.com:gorhom/react-native-bottom-sheet#refactor/convert-variables-to-shared-values
```

**Reply by gorhom** [reply]
here is the pr: https://github.com/gorhom/react-native-bottom-sheet/pull/430

**Comment by gorhom** [comment]
##  `enablePanDownToClose`

This prop will give the ability to pan down the sheet to close, this will also benefit the bottom sheet modal too.

### Testing

```bash
yarn add ssh://git@github.com:gorhom/react-native-bottom-sheet#feature/added-enable-pan-down-to-close
```

**Reply by gorhom** [reply]
pr: https://github.com/gorhom/react-native-bottom-sheet/pull/437

**Comment by raajnadar** [comment]
I found that the bottom sheet is visible for a second when the keyboard toggles.

**Reply by gorhom** [reply]
could you create an issue and i will look into it later 👍

**Comment by darrylyoung** [comment]
This looks great. Thanks for sharing the progress, @gorhom.

For now, should I assume that issues like #52 won't be fixed? It makes sense if you're working on a refactor but any information would be good to help me decide which route to take — use version 2 for now as 3 isn't working or use neither and wait for version 4.

Thanks again and good luck with version 4.

**Reply by gorhom** [reply]
Hi @darrylyoung #52 should been already fixed with the stable release of `v3`.

make sure you follow Reanimated v2 installation instructions https://docs.swmansion.com/react-native-reanimated/docs/next/installation

**Comment by axeldelafosse** [comment]
Hey @gorhom!

Thank you very much for your work.

@nandorojo sent me your [gist to test BottomSheet v4 with RN Modal](https://gist.github.com/gorhom/9cd6155e8469c61d2ae9a815fcc83d2e). It's working fine on iOS but there is a weird behavior on web:

https://user-images.githubusercontent.com/10477267/119509570-0c490c00-bd71-11eb-8b1a-cb5da4a12789.mp4

Any idea why?

**Reply by nandorojo** [reply]
@gorhom Keep in mind that to reproduce this, you need `react-native-web: ^0.15` or higher.

**Reply by gorhom** [reply]
@axeldelafosse, @nandorojo thanks for reporting this issue! i am a bit surprise the library works on web 🤯

I will investigate it later today.

**Reply by nandorojo** [reply]
> i am a bit surprise the library works on web 🤯

@gorhom so am I!

@axeldelafosse and I have some exciting things in the works for universal RN components, and this lib will play an important role for many of them. Thanks for your help!

**Reply by axeldelafosse** [reply]
Yeah! And it works pretty well!

I only needed this patch to make it work:

```
diff --git a/node_modules/@gorhom/bottom-sheet/lib/commonjs/components/touchables/Touchables.android.js.map b/node_modules/@gorhom/bottom-sheet/lib/commonjs/components/touchables/Touchables.android.js.map
deleted file mode 100644
index 1a8f11a..0000000
--- a/node_modules/@gorhom/bottom-sheet/lib/commonjs/components/touchables/Touchables.android.js.map
+++ /dev/null
@@ -1 +0,0 @@
-{"version":3,"sources":["Touchables.android.tsx"],"names":[],"mappings":";;;;;;;;;;;;;;;;;;;;;;;;AAAA","sourcesContent":["export {\n  TouchableOpacity,\n  TouchableHighlight,\n  TouchableWithoutFeedback,\n} from 'react-native-gesture-handler';\n"]}
\ No newline at end of file
diff --git a/node_modules/@gorhom/bottom-sheet/lib/commonjs/components/touchables/Touchables.android.js b/node_modules/@gorhom/bottom-sheet/lib/commonjs/components/touchables/Touchables.js
similarity index 100%
rename from node_modules/@gorhom/bottom-sheet/lib/commonjs/components/touchables/Touchables.android.js
rename to node_modules/@gorhom/bottom-sheet/lib/commonjs/components/touchables/Touchables.js
diff --git a/node_modules/@gorhom/bottom-sheet/lib/commonjs/components/touchables/Touchables.js.map b/node_modules/@gorhom/bottom-sheet/lib/commonjs/components/touchables/Touchables.js.map
new file mode 100644
index 0000000..a16bc78
--- /dev/null
+++ b/node_modules/@gorhom/bottom-sheet/lib/commonjs/components/touchables/Touchables.js.map
@@ -0,0 +1 @@
+{"version":3,"sources":["Touchables.tsx"],"names":[],"mappings":";;;;;;;;;;;;;;;;;;;;;;;;AAAA","sourcesContent":["export {\n  TouchableOpacity,\n  TouchableHighlight,\n  TouchableWithoutFeedback,\n} from 'react-native-gesture-handler';\n"]}
\ No newline at end of file
diff --git a/node_modules/@gorhom/bottom-sheet/lib/module/components/touchables/Touchables.android.js.map b/node_modules/@gorhom/bottom-sheet/lib/module/components/touchables/Touchables.android.js.map
deleted file mode 100644
index 3098e7d..0000000
--- a/node_modules/@gorhom/bottom-sheet/lib/module/components/touchables/Touchables.android.js.map
+++ /dev/null
@@ -1 +0,0 @@
-{"version":3,"sources":["Touchables.android.tsx"],"names":["TouchableOpacity","TouchableHighlight","TouchableWithoutFeedback"],"mappings":"AAAA,SACEA,gBADF,EAEEC,kBAFF,EAGEC,wBAHF,QAIO,8BAJP","sourcesContent":["export {\n  TouchableOpacity,\n  TouchableHighlight,\n  TouchableWithoutFeedback,\n} from 'react-native-gesture-handler';\n"]}
\ No newline at end of file
diff --git a/node_modules/@gorhom/bottom-sheet/lib/module/components/touchables/Touchables.android.js b/node_modules/@gorhom/bottom-sheet/lib/module/components/touchables/Touchables.js
similarity index 70%
rename from node_modules/@gorhom/bottom-sheet/lib/module/components/touchables/Touchables.android.js
rename to node_modules/@gorhom/bottom-sheet/lib/module/components/touchables/Touchables.js
index fc81621..b1edc97 100644
--- a/node_modules/@gorhom/bottom-sheet/lib/module/components/touchables/Touchables.android.js
+++ b/node_modules/@gorhom/bottom-sheet/lib/module/components/touchables/Touchables.js
@@ -1,2 +1,2 @@
 export { TouchableOpacity, TouchableHighlight, TouchableWithoutFeedback } from 'react-native-gesture-handler';
-//# sourceMappingURL=Touchables.android.js.map
\ No newline at end of file
+//# sourceMappingURL=Touchables.js.map
\ No newline at end of file
diff --git a/node_modules/@gorhom/bottom-sheet/lib/module/components/touchables/Touchables.js.map b/node_modules/@gorhom/bottom-sheet/lib/module/components/touchables/Touchables.js.map
new file mode 100644
index 0000000..ac59a11
--- /dev/null
+++ b/node_modules/@gorhom/bottom-sheet/lib/module/components/touchables/Touchables.js.map
@@ -0,0 +1 @@
+{"version":3,"sources":["Touchables.tsx"],"names":["TouchableOpacity","TouchableHighlight","TouchableWithoutFeedback"],"mappings":"AAAA,SACEA,gBADF,EAEEC,kBAFF,EAGEC,wBAHF,QAIO,8BAJP","sourcesContent":["export {\n  TouchableOpacity,\n  TouchableHighlight,\n  TouchableWithoutFeedback,\n} from 'react-native-gesture-handler';\n"]}
\ No newline at end of file
diff --git a/node_modules/@gorhom/bottom-sheet/src/components/touchables/Touchables.android.tsx b/node_modules/@gorhom/bottom-sheet/src/components/touchables/Touchables.tsx
similarity index 100%
rename from node_modules/@gorhom/bottom-sheet/src/components/touchables/Touchables.android.tsx
rename to node_modules/@gorhom/bottom-sheet/src/components/touchables/Touchables.tsx
```

Let me know if you need help while setting up your web app. I recommend using Expo + Next.js :)

**Reply by nandorojo** [reply]
This patch didn't solve the issue in your video, right? Is it just what's needed for web compatibility in general?

**Reply by gorhom** [reply]
@nandorojo , @axeldelafosse  could you guys help me to submit a pr to add `react-native-web` to `/examples` on `v4` branch ?

i would really appreciate it 👏

**Reply by axeldelafosse** [reply]
Yes, the patch is just for web support.

OK I'll add an example tomorrow!

**Reply by axeldelafosse** [reply]
Here we go: https://github.com/gorhom/react-native-bottom-sheet/pull/460

**Reply by axeldelafosse** [reply]
Also I'm not sure if I was clear earlier so just wanted to note that in this example and in the video above, the issue is because we use a React Native `<Modal />` -- it works correctly with `<BottomSheetModal />` for example.

**Reply by nandorojo** [reply]
@axeldelafosse thanks for the PR!

Yeah @gorhom it actually already works on web with the BottomSheetModal it seems. It's only the RN modal that experiences this.

**Reply by axeldelafosse** [reply]
No worries! It was long overdue.

I'll try to make it work with `ShowcaseApp` if I find some time this weekend.

**Reply by gorhom** [reply]
Thanks @axeldelafosse for submitting the pr, tomorrow i'll review it and look into the web issue 👏

**Reply by nandorojo** [reply]
Thanks for the quick work here guys

**Reply by axeldelafosse** [reply]
Hey guys! Going back to this one. Here is a repo featuring this bug: https://github.com/axeldelafosse/bottom-sheet-example. Also, I realized that the modal isn't moving on Android 🤷‍♂️

**Comment by gorhom** [comment]
## Added BottomSheetFooter 

This component will be placed at the bottom of the sheet all the time and will react to keyboard appearnce.


### Props

`appearanceBehavior`: Appearance behavior when sheet is below starts to cut off footer, you can combine many behaviors together.
* `none`: do nothing.
* `fade`: fade in and out.
* `scale`: scale up and down.
* `slide`: slide up and down.

https://user-images.githubusercontent.com/4061838/119574270-610a7800-bdad-11eb-978e-9ea0971d197d.mp4

## Testing
```bash
yarn add @gorhom/bottom-sheet@4.0.0-alpha.4
```

**Comment by gorhom** [comment]
## Prevent resetting the scrollable position when dragging by the handle

more info here #461

https://user-images.githubusercontent.com/4061838/120119972-765b1a00-c192-11eb-8531-edb9ebb652a2.mov

## Testing

```bash
yarn add @gorhom/bottom-sheet@4.0.0-alpha.7
```

**Reply by nandorojo** [reply]
@gorhom where can I see the code for this search list UI?

**Reply by gorhom** [reply]
here you go https://github.com/gorhom/react-native-bottom-sheet/blob/v4/example/src/components/searchHandle/SearchHandle.tsx

**Comment by gorhom** [comment]
## Detached Bottom Sheet

This will allow user to create a detached bottom sheet with less effort.

https://user-images.githubusercontent.com/4061838/121781720-068b5d00-cb9e-11eb-953c-488dc313833b.mp4

## Testing

```bash
yarn add @gorhom/bottom-sheet@4.0.0-alpha.12
```

**Reply by nandorojo** [reply]
So cool!

**Comment by alessiocancian** [comment]
I think this feature would be nice to have, especially along with web support:
https://github.com/gorhom/react-native-bottom-sheet/discussions/234

I could help with the implementation if needed

**Reply by gorhom** [reply]
@alessiocancian i think it will be doable now since we have detached bottom sheet implementation in place , would you mind have a look at it and see if its doable ? thanks alot

**Reply by alessiocancian** [reply]
@gorhom  I'm on it.
I looked at the BottomSheet implementation and I think I need to write a different component for this feature by using some of the internal components such as BottomSheetBackdrop, BottomSheetContainer and BottomSheetBackgroundContainer.
A wrapper will choose the component to render between BottomSheet and ContainedLayout. I think it should be based on a prop of the BottomSheet and not just calculated internally based on the screen size, but this could be another option. Like a `displayMode` prop with `default | contained | auto` as options.

Then we need to find a way to translate the snap points to the modal display height, we could choose the greatest one or add a different prop to handle modal height...

Another feature to consider is the modal header. Usually when presenting a modal instead of the bottom-sheet I like to show an header with a title and the close button. I'd like to add a prop to the BottomSheet, like `modalTitle` or `containedLayoutTitle` to show this header automatically only for the contained layout.

**Comment by gorbypark** [comment]
I'd love it if we could get the onScroll event working for BottomSheetScrollView and related components.  I have a few animations that need to run when the y contentOffset !== 0 but currently can't because there's no way (that I know of) to get that information. https://github.com/gorhom/react-native-bottom-sheet/issues/488

**Comment by nandorojo** [comment]
I've been getting this issue on v4, a bit hard to reproduce but notice it only happens like 1/10 times (midway through the video it shoots up.)

I'm using the sheet inside of a `Modal`. Let me know if you need me to share the code.

https://user-images.githubusercontent.com/13172299/123491539-ea97aa80-d5e4-11eb-9b71-ab1236e3b19d.MOV

**Comment by Vednus** [comment]
Not sure if this is the place to write this, but it would be nice to have the option to use it in a more declarative way so I can just set the `index` prop and whenever that changes, it moves to that index. I often just handle the index in a state property and then use a useEffect to `snapToIndex` when that index changes. It's fine for now, but might be nice if index behaved like that.

**Reply by gorhom** [reply]
this will be added with alpha 21 👏

**Comment by elgorditosalsero** [comment]
Hi @gorhom, quick question...  😄 

As in V3 there is the [dismiss by name/key](https://gorhom.github.io/react-native-bottom-sheet/modal/hooks#dismiss) functionality, is possible also add the `expand` by name/key in v4?

**Comment by skdev24** [comment]
Hey @gorhom 

I was testing `BottomSheetModal` on 4.0.0-alpha.17, seems like `onDismiss` method is not getting called, to which I did little testing found out that `handleOnChange` method from `BottomSheet` was not getting called because of inside `useAnimatedReaction` condition was false every time.

https://github.com/gorhom/react-native-bottom-sheet/blob/7139e08e3aac12c64d6a05be081cb3212f5674fe/src/components/bottomSheet/BottomSheet.tsx#L1598 

But as I seen the value of both condition I changed with equals to sign which worked.

**Reply by gorhom** [reply]
thanks for spotting it, it should be fixed with `alpha.18`

**Comment by gorbypark** [comment]
Hey @gorhom, it looks like from 4.0.0-alpha.16 BottomSheetScrollView doesn't scroll anymore.  I've tried on alpha.16/17/18, downgrading to alpha.15 and everything works again.  I have a stack navigator inside of the bottom sheet.

```js
        <BottomSheet
          ref={bottomSheetRef}
          snapPoints={snapPoints}
          animatedPosition={currentPosition}
          onChange={index => {
            index === 1 ? handleBottomSheetOpen() : handleBottomSheetClose();
          }}
          backdropComponent={props => (
            <BottomSheetBackdrop {...props} pressBehavior="collapse" />
          )}
          handleComponent={props => <BottomSheetHandle {...props} />}
          backgroundComponent={props => <BottomSheetBackground {...props} />}>
          <BottomSheetStack.Navigator
            screenOptions={{
              headerTintColor: 'rgba(185,38,50,1)',
              headerTitleStyle: {color: 'black'},
              headerStyle: {
                height: 46, // This matches the height of the bottom sheet when closed
              },
            }}>
            <BottomSheetStack.Screen
              name="BottomSheetHome"
              options={{headerShown: false}}
              component={BottomSheetHomeScreen}
            />
            <BottomSheetStack.Screen
              name="Frequently"
              options={{headerStatusBarHeight: 0}}
              component={FrequentlyScreen}
            />
            <BottomSheetStack.Screen
              name="Favourite"
              options={{headerStatusBarHeight: 0}}
              component={FavouriteScreen}
            />
            <BottomSheetStack.Screen
              name="Recent"
              options={{headerStatusBarHeight: 0}}
              component={RecentScreen}
            />
            <BottomSheetStack.Screen
              name="Featured"
              options={{headerStatusBarHeight: 0}}
              component={FeaturedScreen}
            />
          </BottomSheetStack.Navigator>
        </BottomSheet>

```
and then this is my BottomSheetHomeScreen component with the BottomSheetScrollView that has stopped working.
```js
    <BottomSheetView style={styles.contentContainer}>
      <SearchBox />
      <Divider />
      <BottomSheetScrollView>
        <Frequent />
        <Favourites />
        <Recents />
        <Featured />
        <Frequent />
        <Favourites />
        <Recents />
        <Featured />
      </BottomSheetScrollView>
    </BottomSheetView>
```
 BottomSheetScrollView has stopped scrolling, but still can close the bottom sheet.  Switching out for ScrollView from both `react-native` and `react-native-gesture-handler` allows scrolling, but doesn't allow closing of the bottom sheet (obviously).  Here's the unchanged component in question, first one is on alpha.18 and the second one is on alpha.15

*edit*
After taking a look at the commit for alpha.16, It appears that setting `enableOverDrag={false}` on BottomSheet fixes my issue.  Not quite sure why 
```js
          if (
            (type === GESTURE_SOURCE.HANDLE ||
              animatedScrollableType.value === SCROLLABLE_TYPE.VIEW) &&
            draggedPosition < highestSnapPoint
          )
``` 
is causing the issue while overscrolling on just the handle does not.  Maybe this block is gobbling up the gesture event and it's not making it's way to the BottomSheetScrollView component?
 

https://user-images.githubusercontent.com/19569469/124226109-d7b33780-db08-11eb-9fa7-c07cf83aea8a.mp4

https://user-images.githubusercontent.com/19569469/124226314-34aeed80-db09-11eb-9a4a-bef021796bc2.mp4

**Reply by gorhom** [reply]
you do not need to wrap your `BottomSheetHomeScreen` with `BottomSheetView`

**Comment by pouyarezvani** [comment]
@gorhom when can we expect v4 to be fully released?

**Reply by gorhom** [reply]
i am working on finalising it , currently i would appreciate if you could test `v4` and let me know if you find any issues 

here are the remaining tasks: 
- Insure all new features are working as expected.
- Move `v4` branch to `master` 
- Complete and move website to the new `master`

**Comment by nmuranovas** [comment]
Hello 👋

It would be really nice to have complete control over the accessibility props (accessible, accessibilityLabel, accessibilityRole, accessibilityHint, accessibilityState) on the BottomSheet as well as it's sub-components (handle, backdrop) since it is currently not possible to do localisation.

In my team we have done some research on how accessibility is handled on different apps, platforms and it seems like everyone does it completely differently (e.g. some do not make the backdrop accessible, others do not even make the handle accessible and basically anything in between).

Are there any plans in introducing customisable accessibility in the v4 release?

**Reply by nmuranovas** [reply]
In the v4 release or either by contributing to the current version?

**Reply by gorhom** [reply]
Hi @nmuranovas, i will priorities `accessibility` after this major release, please feel free to open a new discussion regarding what's needed or possible implementation :)

**Comment by ShivamJoker** [comment]
@gorhom there seems to be issue with the backdrop component with `alpha.17`
When the snap is at 1st point it doesn't show backdrop but with the 2nd one it shows

https://user-images.githubusercontent.com/23727670/126894240-90c08ee2-062d-472d-abf9-cf0beb5fd3f9.mov

I am simply taking the bottom sheet backdrop from the library
Also the FlatList disappears when SearchInput gets focused

**Reply by gorhom** [reply]
@ShivamJoker you can override `BottomSheetBackdrop` props to control your expected behaviour.

```tsx
<BottomSheetBackdrop appearsOnIndex={1} disappearsOnIndex={-1} />
```

**Comment by ajonikas** [comment]
The refresh control implementation works on Expo 41 but seems to break on Expo 42. Small snack to replicate: https://snack.expo.dev/4IW1JzpPC

**Reply by gorhom** [reply]
@ajonikas you need to pass `onRefresh` to the scrollable in order to get it working 👍

**Reply by ajonikas** [reply]
Ah I see! Thank you

**Comment by thebiltheory** [comment]
Good Afternoon @gorhom 

I'm am continuing the convo here instead of Twitter.

Here is some feedback:

### The snack has some issues. 
![Screenshot 2021-08-11 at 14 12 45](https://user-images.githubusercontent.com/2975322/129026628-addbc7c6-6225-4544-9f39-e9c3c08d70a1.png)

### React Navigation integration

When trying to implement a navigator (as below and according to your examples) nothing happens. 

The code below is the way I implemented it. 

```javascript

  const {
    animatedHandleHeight,
    animatedSnapPoints,
    animatedContentHeight,
    handleContentLayout,
  } = useBottomSheetDynamicSnapPoints(["CONTENT_HEIGHT"]);


<NativeViewGestureHandler disallowInterruption={true}>
      <View flex={1}>
        <BottomSheetModal
          name="Select"
          ref={newBookingModal.ref}
          onDismiss={resetCarForm}
          snapPoints={animatedSnapPoints}
          handleHeight={animatedHandleHeight}
          contentHeight={animatedContentHeight}
          enablePanDownToClose={true}
          keyboardBehavior="interactive"
          android_keyboardInputMode="adjustPan"
        >
          <BottomSheetScrollView
            onLayout={handleContentLayout}
            style={{
              flex: 1,
              backgroundColor: "red",
            }}
          >
            <NewBookingNavigator />
          </BottomSheetScrollView>
        </BottomSheetModal>
      </View>
    </NativeViewGestureHandler>
```


`NewBookingNavigator.ts`
```javascript
 <NavigationContainer independent={true}>
      <Stack.Navigator
        screenOptions={screenOptions}
        initialRouteName="ScreenOne"
      >
        <Stack.Screen
          name="ScreenOne"
          options={{ title: "Screen One", headerShown: false }}
          component={ScreenOne}
        />
        <Stack.Screen
          name="ScreenTwo"
          options={{ title: "Screen Two", headerShown: false }}
          component={ScreenTwo}
        />
        ...
      </Stack.Navigator>
    </NavigationContainer>
```

**Reply by thebiltheory** [reply]
Hey @gorhom, 
Any news on this?
I have a demo on Sunday  (It's the first day of the week in the UAE), would have been amazing to have this working.

Not even sure if this is a bug or a wrong implementation of the new API.

**Reply by gorhom** [reply]
I will look into it later today

**Reply by thebiltheory** [reply]
Looking forward to it! Lemme know if I can be of any help!

**Reply by gorhom** [reply]
@thebiltheory this issue only happens when testing on Expo Snack web, it should be fine when using on Android or iOS

**Reply by thebiltheory** [reply]
You mean the navigator? It's not working on iOS nor android. I'll try again tomorrow and try to give you more context.

**Reply by thebiltheory** [reply]
Hey @gorhom,

I've made a snack where Navigators Screen isn't working with Dynamic Snap Points.
https://snack.expo.dev/@thebiltheory/bottom-sheet-v4-with-react-navigation

For some reason, the header is showing in the snack but not in my implementation.

**Reply by bohemima** [reply]
I'm seeing the same behaviour as @thebiltheory, web, iOS & Android - having a working auto-sizing bottomsheet with navigation would be awesome and replace my own hacky solution

**Reply by thebiltheory** [reply]
What's your hacky solution @bohemima ?
I'd love to make this work somehow without have to snap to a specific snap point index on every `navigate` action.

**Reply by bohemima** [reply]
There's a hacky PoC on my profile, react-navigation-modal-as-bottomsheet-poc - it's not using a 'real' bottom sheet at all, just views and some transparency. No keyboard handling, no swipe to dismiss.

I'll see if I get some time to deep dive into @gorhom's code and see if we can't solve this somehow..

**Reply by thebiltheory** [reply]
@bohemima

I've implemented a hack that works.
Checkout the implementation.
https://github.com/gorhom/react-native-bottom-sheet/issues/601

**Reply by bohemima** [reply]
Good stuff, could use context to share the method instead, or maybe there's issues when using a modal?

**Reply by thebiltheory** [reply]
@bohemima at this level I think it's alright to pass the method as a prop. The only thing is that React Navigation is gonna complain about non-serializable data (the method). you can suppress that with LogBox. 

Feel free to suggest improvements!

**Comment by nandorojo** [comment]
I'm seeing this error after upgrading from alpha 13 → 28.

![image](https://user-images.githubusercontent.com/13172299/129950907-35d8e090-c26c-4107-bd5d-b859193227db.png)

I also noticed that the footer no longer has the `appearanceBehavior` prop. Can this still be set?

Is the newly required `animatedFooterPosition` prop on the footer just an animated value?

**Reply by gorhom** [reply]
unfortunately i had to update the Footer api, it was super limited to pre-configure simple animation appearance, so i had to make it more generic to enable users to create their own custom animation.

look the at new custom footer component https://github.com/gorhom/react-native-bottom-sheet/blob/master/example/src/components/customFooter/CustomFooter.tsx

**Comment by moyolvera** [comment]
Hi, i miss some updates and just got the new alpha 30, however I'm getting that the use of 0 as a Snap Point  is now being disabled and a red logbox window appears, it says that i need to add the enablePanDownToClose prop which i have, what i want to do is to have the bottomsheet closed by default, it's possible to do with some new prop or something?

**Reply by gorhom** [reply]
yeah, you could set `index={-1}`

**Reply by moyolvera** [reply]
thank you sir

**Comment by eden-lane** [comment]
Hi everyone!
I have a problem with installation, maybe someone knows how to fix this.
I've installed `@gorhom/bottom-sheet@4.0.0-alpha.30`,
I've installed `react-native-gesture-handler` and `react-native-reanimated`. 
I've completed all steps of installation from https://docs.swmansion.com/react-native-reanimated/docs/installation/

but I'm still getting this error on iOS:
![image](https://user-images.githubusercontent.com/2963477/130615003-09fe6586-452d-4988-bb06-e57a50b8a503.png)

**Reply by nandorojo** [reply]
I suggest opening this as an issue, unless it worked before v4 and then stopped working.

**Comment by alexco2** [comment]
Hi everyone,
I upgraded from `v3` to `v4`. So far everything works well. Especially the `footerComponent` is a huge benefit for me. I´m using a `StackNavigator` in my `BottomModal` though and want to render a different `footerComponent` based on the `navigationState` the Stack has. Do you know if there is a possibility to control the `footerComponent` prop from a nested stack screen?
Thank you for your help! :)

**Reply by gorhom** [reply]
you could copy `animatedFooterPosition`  from `BottomSheetFooterContainer` and add it to your custom footer per screen/stack 👍

**Reply by alexco2** [reply]
Thank you! I solved it by passing all of the state down the tree. Ideally, something like react navigations setOptions (headerRight, headerLeft etc.) would have been helpfull, but I think that really is just a niche problem :)

**Comment by theMasix** [comment]
I'm waiting for stable `v4` and its documentation for weeks...

**Reply by gorhom** [reply]
documentation is up https://gorhom.github.io/react-native-bottom-sheet , i am just finalising the stable release and it would be on Monday 🤞

**Comment by alexco2** [comment]
@gorhom 
Hello, I come across one issue, but unfortunatly I am unable to reproduce it in a snack. When I unmount the screen which has a `BottomSheetModal`, the app crashes. This just happens when the modal is closed (e.g. by tapping the backdrop). If it is "manually" dismissed, the crash does not appear. I tried to narrow the problem down, but even after stripping the modal of all content, the issue still persists. Is it perhaps a problem that I am using one `BottomSheetModalProvider` at the top level for all my `BottomSheetModal` components? Do you have an idea what could cause the problem?

The only error log I can provide is one from crashlytics:

```
Fatal Exception: java.lang.IllegalArgumentException: Scrapped or attached views may not be recycled. isScrap:false isAttached:true b.v.b.f$m{b086991 VFED..... ......ID 0,0-1080,2096 #1a}, adapter:com.reactnativepagerview.b@9e550f6, layout:b.v.b.f$h@9929cf7, context:com.facebook.react.uimanager.l0@7d7890e
       at androidx.recyclerview.widget.RecyclerView$v.C(:282)
       at androidx.recyclerview.widget.RecyclerView.removeAnimatingView(:22)
       at androidx.recyclerview.widget.RecyclerView$m.a(:27)
       at androidx.recyclerview.widget.RecyclerView$l.h(:7)
       at androidx.recyclerview.widget.n.G(:3)
       at androidx.recyclerview.widget.c$d.onAnimationEnd(:17)
       at android.view.ViewPropertyAnimator$AnimatorEventListener.onAnimationEnd(ViewPropertyAnimator.java:1111)
       at android.animation.Animator$AnimatorListener.onAnimationEnd(Animator.java:554)
       at android.animation.ValueAnimator.endAnimation(ValueAnimator.java:1250)
       at android.animation.ValueAnimator.doAnimationFrame(ValueAnimator.java:1492)
       at android.animation.AnimationHandler.doAnimationFrame(AnimationHandler.java:146)
       at android.animation.AnimationHandler.access$100(AnimationHandler.java:37)
       at android.animation.AnimationHandler$1.doFrame(AnimationHandler.java:54)
       at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1008)
       at android.view.Choreographer.doCallbacks(Choreographer.java:809)
       at android.view.Choreographer.doFrame(Choreographer.java:740)
       at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:995)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:246)
       at android.app.ActivityThread.main(ActivityThread.java:8512)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```

**Reply by alexco2** [reply]
I found this issue.
https://github.com/callstack/react-native-pager-view/issues/173
When switching the parent `nativeStackNavigator` to a normal `StackNavigator` the issue is resolved. For performance reasons I would rather use the native version. Is there a way to make sure the `BottomSheetModal` is alwas dismissed?

**Reply by gorhom** [reply]
could you provide a snack using the `v4` template ?

**Reply by alexco2** [reply]
I would like to, but I´m getting the repeated error that expo failed to resolved dependencies because the network request failed for `react-native-pager-view`.
https://snack.expo.dev/@expoco2/fatal-crash-bottomsheetmodal-v4-with-pagerview
In theory, the bug should be reproducable by going back from a `NativeStack` screen which has a `PagerView` (`react-native-pager-view`) or `TabView` (`react-native-tab-view`) and a not unmounted `BottomSheetModal` (not dismissed, but e.g. closed).
I will check over the day if the expo issue might be resolved.
